### PR TITLE
Update JTS and Spatial4j Dependencies

### PR DESCRIPTION
**Pull request by worace**
I'm trying to resuscitate this library a bit so I can start using it more for various projects.

One of the first things I wanted to tackle was getting our versions of JTS and Spatial4J updated since they're several years old at this point.

During the process of these updates I ran into a few corner-cases around handling geometries that cross the international dateline and had to resolve those.

I think I have it working now at least as well as it was before, and I was able to fix a couple geohash edge cases along the way by simplifying our intersecting-geohashes implementation a bit as well.

Also took the opportunity to add a simple IO namespace with some read/write utility functions.

### CODE REVIEW

File: src/geo/geohash.clj [hunk 1]
```diff
@@ -33,21 +36,11 @@
                               (.getMinLat box)
                               (.getMaxLat box)))))
 
-(defn northern-neighbor
```
Comment (chen-factual) [inline src/geo/geohash.clj]: Type hinting the above GeoHash to shape might be good. I imagine this maybe used rather often.

Actually generally speaking when dealing with a lot of Java like this, it's good to turn on reflection warnings and see where you're taking that hit.

File: src/geo/geohash.clj [hunk 2]
```diff
@@ -192,82 +188,38 @@
...
-  resolution which intersect that shape."
-  [shape precision ^GeoHash geohash]
-  (let [relationship (relate shape geohash)]
-    (cond
-      (= relationship :contains)
-      (do
-;        (print "X")
-        (subdivide geohash precision))
+(defn children [gh]
+  (->> gh string GeohashUtils/getSubGeohashes (map geohash)))
```
Comment (worace) [inline src/geo/geohash.clj]: I think there is a better way to do this actually

File: src/geo/geohash.clj [hunk 3]
```diff
@@ -192,82 +188,38 @@
...
-;              (height shape) " x " (width shape))
-     (geohashes-intersecting shape precision initial-precision)))
-  ([shape precision initial-precision]
-   (mapcat (partial geohashes-intersecting-recursive shape precision)
-           (geohashes-intersecting-rings shape initial-precision))))
+  ([shape desired-level] (geohashes-intersecting shape desired-level desired-level))
+  ([shape min-level max-level]
+   (loop [matches #{}
+          queue (list (geohash ""))]
+     (if (empty? queue) matches
```
Comment (chen-factual) [inline src/geo/geohash.clj]: am I reading this right? I only see recurs, I don't actually see something that returns.

File: src/geo/geohash.clj [hunk 4]
```diff
@@ -192,82 +188,38 @@
...
+         (let [current (first queue)
+               level (significant-bits current)]
+
+           (if (and (<= level max-level) (intersects? shape current))
+             (if (= level max-level) (recur (conj matches current)
```
Comment (chen-factual) [inline src/geo/geohash.clj]: mixing same-line and multi-line if-elses is very hard to read, you generally want to use one or the other. I.e. do either

```clojure
(if my-cond? 1 0)
```
or 
```clojure
(if my-cond?
  1
  0)
```
But mixing is very confusing:
```clojure
(if my-cond? 1
  0)
```

File: src/geo/geohash.clj [hunk 5]
```diff
@@ -192,82 +188,38 @@
...
+                                            (rest queue))
+                 (if (>= level min-level)
```
Comment (chen-factual) [inline src/geo/geohash.clj]: There's a lot of recurs here. One way to maybe clean this up is to do something like
```clojure
(let [new-matches (update-matches matches current level min-level max-level)
      new-queue (update-queue queue current level min-level max-level)]
  (recur new-matches new-queue))
```

Also there's a ton of seq operations here, and `into` can potentially be expensive, so look out for that if you see performance issues.

File: src/geo/spatial.clj [hunk 6]
```diff
@@ -93,7 +100,16 @@
 
   com.vividsolutions.jts.geom.Geometry
   (to-shape [this]
-           (JtsGeometry. this jts-earth true)))
+    ;; Cloning geometries that cross dateline to workaround
+    ;; spatial4j / jts conversion issue: https://github.com/locationtech/spatial4j/issues/150
+    (let [geom (if (crosses-dateline? this)
+                 (.clone this)
+                 this)]
```
Comment (chen-factual) [inline src/geo/spatial.clj]: A way to make code self-documenting is to define named variables for the literals you'd use. So here, you can add to your `let`:

```clojure
(let [dateline-check? true
      allow-multi-overlap? false]
  ...
)
```

**Comment by worace** [comment]
@chen-factual awesome thanks for those comments; will do an update based on this in a bit.

I've never actually used the reflection warnings feature; will try that out.

**Pull request merged**